google/cloud/aiplatform/models.py

```diff
@@ -1259,8 +1299,467 @@ def explain(
+                Endpoint will be undeployed first. Default is False.
+            sync (bool):
+                Whether to execute this method synchronously. If False, this method
+                will be executed in concurrent Future and any downstream object will
+                be immediately returned and synced when the Future has completed.
+
+        Raises:
+            FailedPrecondition: If models are deployed on this Endpoint and force = False.
+        """
+        if force:
+            self.undeploy_all(sync=sync)
+
+        super().delete(sync=sync)
+
+
+class PrivateEndpoint(Endpoint):
+    """
+    Represents a Vertex AI private Endpoint resource.
+
+    Read more [about private endpoints in the documentation.](https://cloud.google.com/vertex-ai/docs/predictions/using-private-endpoints)
+    """
+
+    def __init__(
+        self,
+        endpoint_name: str,
+        project: Optional[str] = None,
+        location: Optional[str] = None,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ):
+        """Retrieves a private Endpoint resource.
+
+        Example usage:
+
+            my_private_endpoint = aiplatform.PrivateEndpoint(
+                endpoint_name="projects/123/locations/us-central1/endpoints/my_endpoint_id"
+            )
+
+            or (when project and location are initialized)
+
+            my_private_endpoint = aiplatform.PrivateEndpoint(
+                endpoint_name="my_endpoint_id"
+            )
+
+        Args:
+            endpoint_name (str):
+                Required. A fully-qualified endpoint resource name or endpoint ID.
+                Example: "projects/123/locations/us-central1/endpoints/my_endpoint_id" or
+                "my_endpoint_id" when project and location are initialized or passed.
+            project (str):
+                Optional. Project to retrieve endpoint from. If not set, project
+                set in aiplatform.init will be used.
+            location (str):
+                Optional. Location to retrieve endpoint from. If not set, location
+                set in aiplatform.init will be used.
+            credentials (auth_credentials.Credentials):
+                Optional. Custom credentials to use to upload this model. Overrides
+                credentials set in aiplatform.init.
+        """
+        super().__init__(
+            endpoint_name=endpoint_name,
+            project=project,
+            location=location,
+            credentials=credentials,
+        )
+
+        self._custom_predict_uri = None
+        self._custom_explain_uri = None
+        self._custom_health_uri = None
+
+        self._http_client = urllib3.PoolManager()
+
+    @property
+    def predict_http_uri(self) -> Optional[str]:
+        """Http(s) path to send prediction requests to, used when calling `PrivateEndpoint.predict()`"""
+        if not self._gca_resource.deployed_models:
+            raise RuntimeError(
```

Comment:
So the `predict`, `explain` and `health_check` methods all use their corresponding uris. I initially though the same way (putting the exception in the method itself), but if a user calls on the uris without deploying a model, the line
```
self._gca_resource.deployed_models[0].private_endpoints.explain_http_uri
```
in the uri methods will return an index out of range error, instead of the RuntimeError. The way it's written now, both the uri methods and the `predict`, `explain` and `health_check` methods will raise the RuntimeError.

Let me know what approach we should take. Thanks

Reply:
hey Sasha, I added `None` to the return value of the public properties and moved the exceptions to the methods